WIP: Cpp Export

patrick-luethi opened the pull request:
- Added more logic to forward declaration printing
  - Now groups forward declarations from the same namespace
  - Forward declarations are put into previous unit's namespace declarations if applicable
- Split method in smaller methods to increase readability
- Added comments
  <a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/dimitar-asenov/Envision/pull/337%23issuecomment-189319576%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/337%23issuecomment-189319576%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22done%22%2C%20%22created_at%22%3A%20%222016-02-26T15%3A18%3A21Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6049349%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/patrick-luethi%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%7D%7D'></a>
  <a href='https://www.codereviewhub.com/'><img src='http://www.codereviewhub.com/site/github-bar.png' height=40></a>
- [ ] <a href='#crh-comment-General Comment'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/337#issuecomment-189319576'>General Comment</a></b>
- <a href='https://github.com/patrick-luethi'><img border=0 src='https://avatars.githubusercontent.com/u/6049349?v=3' height=16 width=16'></a> done

<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/337?mark_as_completed=1'><img src='http://www.codereviewhub.com/site/github-mark-as-completed.png' height=26></a>&nbsp;<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/337?approve=1'><img src='http://www.codereviewhub.com/site/github-approve.png' height=26></a>&nbsp;<a href='https://github.com/dimitar-asenov/Envision/pull/337'><img src='http://www.codereviewhub.com/site/github-refresh.png' height=26></a>
<a href='#crh-end'></a>

patrick-luethi committed:
refactor code composite export

- Added more logic to forward declaration printing
	- Now groups forward declarations from the same namespace
	- Forward declarations are put into previous unit's namespace declarations if applicable
- Split method in smaller methods to increase readability
- Added comments

dimitar-asenov commented:
This pull request is now managed with CodeReviewHub. <a href='https://github.com/dimitar-asenov/Envision/pull/337'>Refresh</a>.<a href='#crh-update-none'></a>

patrick-luethi committed:
change forward declaration export

- Fixed an issue with circular dependencies not being handled correctly
- Every code unit can have their own forward declarations (not on namespace block basis)
- Forward declarations are pushed up to previous units in the same namespace
- Use index of units list to reduce the number of necessary datastructures

patrick-luethi committed:
move comment

patrick-luethi committed:
re-purpose method to separate calculation and meaning of usage

patrick-luethi committed:
remove unnecessary condition

patrick-luethi committed:
rename method argument

patrick-luethi committed:
rename method

patrick-luethi committed:
add comments

patrick-luethi commented:
done
